File: Algorithms/src/ywy/java/level1/UnfinishedPlayer.java
package ywy.java.level1;

import java.util.HashMap;
import java.util.Map;

public class UnfinishedPlayer {
	public String solution(final String[] participants, final String[] completions) {
		final Map<String, Integer> participantList = getParticipantList(participants);
		String completion = null;
		int index = 0;
		int range = completions.length;
		Integer temp = null;
		
		do {
			completion = completions[index];
			temp = participantList.get(completion);
			if (temp == null) {
				return completion;
			}
			temp--;
			if (temp == 0) {
				participantList.remove(completion);
				continue;
			}
			participantList.put(completion, temp); 
		} while((++index) < range);
		return participantList.keySet().iterator().next();
	}
	
	private Map<String, Integer> getParticipantList(final String[] participants) {
		final Map<String, Integer> participantList = new HashMap<String, Integer>();
		int index = 0;
		int range = participants.length;
		String participant = null;
		do {
			participant = participants[index];
			participantList.put(participant, participantList.getOrDefault(participant, 0) + 1);
		} while((++index) < range);
		return participantList;
	}
}

File: Ywy/JavaScript/level1/UnfinishedPlayer.js
const solution = (participants, completions) => {
    const {participantObj, completionObj} = getSettingObjs(participants, completions);
    const keys = Object.keys(participantObj);
    let key;
    let index = 0;
    const range = keys.length;
    let completionCount = 0;
    let participantCount = 0;
    do {
        key = keys[index];
        completionCount = completionObj[key];
        if (completionCount === undefined) {
            return key;
        } 
        participantCount = participantObj[key];
        completionObj[key] = --completionCount;
        participantObj[key] = --participantCount;
        if (participantCount - completionCount === 1) {
            return key;
        }
    } while((++index) < range);
    return [participantObj, completionObj]
};

const getSettingObjs = (participants, completions) => {
    const participantObj = {};
    const completionObj = {};
    const range = participants.length - 1;
    let index = 0;
    let key = null;
    participantObj[participants[range]] = 1;
    do {
        key = participants[index];
        participantObj[key] = !participantObj[key] ? 1 : participantObj[key] + 1;
        key = completions[index];
        completionObj[key] = !completionObj[key] ? 1 : completionObj[key] + 1;
    } while((++index) < range);
    return {participantObj: participantObj, completionObj: completionObj};
};